### ISSUE TITLE
how to output the val loss in mmsegmentation?

### ISSUE THREAD

**Issue by SWJTUJoeGit** [rubeea/pl_mmseg_point_rend#1]
I am sorry to bother you!At the mmsegmentaion project,I am confused about the output of val loss,I also changed  the the workflow to [('train', 1), ('val', 1)],but it didn't work,so could you please tell me how you solve it?I am looking forward to hearing from your early reply.

**Comment by rubeea** [comment]
Hi @SWJTUJoeGit no worries. I used the workflow [('train', 1)] with validate="True" option so the output parameters (Accuracy, IoU or Dice) are calculated on the validation dataset (and not on the training dataset). Let me know if you need any other help.

**Comment by SWJTUJoeGit** [comment]
I am so sorry to  reply such late ,thank you so much for your advise.but I am still confused about how to output val loss and train loss at same time.and if I set the parameters like this,the output loss shown in the tensorboard was calculated in the validation dataset?I want to know about the loss on the train and val dataset so that I can figure out my model training well or not. Looking forward your early reply,thank you a lot.

**Comment by rubeea** [comment]
> I am so sorry to reply such late ,thank you so much for your advise.but I am still confused about how to output val loss and train loss at same time.and if I set the parameters like this,the output loss shown in the tensorboard was calculated in the validation dataset?I want to know about the loss on the train and val dataset so that I can figure out my model training well or not. Looking forward your early reply,thank you a lot.

Hi @SWJTUJoeGit,
Yes, if you set the workflow to [('train', 1)] with validate="True" the losses shown in the Tensorboard are calculated on the validation dataset. For displaying the losses on both the train and val dataset in Tensorboard set the workflow to [('train', 1),('val',1)] with validate="False" and we need the key **'log_vars'** in the output dictionary. This key exists in train output (in def train_step), but not in val output. That is why the val loss is not shown in Tensorboard, I suppose.
So, I simply mimic the def train_step. I added the following after output = self(**data_batch, **kwargs) in def val_step in file:
[https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/segmentors/base.py#L162](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/segmentors/base.py#L162)

   ```
 loss, log_vars = self._parse_losses(output)

    import collections
    log_vars_val = collections.OrderedDict()
    for k,v in log_vars.items():
        new_key = 'val_' + k
        log_vars_val[new_key] = v

    output = dict(
        loss=loss,
        log_vars=log_vars_val,
        num_samples=len(data_batch['img'].data))
```
I slightly changed the name by adding prefix 'val_' in the keys, otherwise I think the val loss is not distinguished from train loss in the tensorboard. In my case, this workaround worked and the val loss is shown on the Tensorboard. (One unsatisfactory point is that the val loss is shown in 'train' tab... This is ugly but is not a problem practically.)
I hope this helps.

**Comment by SWJTUJoeGit** [comment]
I am so glad you can reply,and thanks a lot for your great suggestion for my project.Actually I followed your idea to change the code in order to output the loss between val and train.But when I show the val loss and train loss in the fig.I got confused about the result.The first fig. is the output loss that I followed your way.
![fig](https://user-images.githubusercontent.com/57290520/114273673-96e9dc00-9a4d-11eb-95d2-6627e1a50c2d.jpg)
and I also try to run the code twice by only change the workflow,the result seems much normal.
![kkk](https://user-images.githubusercontent.com/57290520/114273820-2c856b80-9a4e-11eb-8e6f-c870ef6478be.jpg)
so I am sure about the loss of the val_step output is correct.Looking forward to your replay.

**Comment by rubeea** [comment]
> I am so glad you can reply,and thanks a lot for your great suggestion for my project.Actually I followed your idea to change the code in order to output the loss between val and train.But when I show the val loss and train loss in the fig.I got confused about the result.The first fig. is the output loss that I followed your way.
> ![fig](https://user-images.githubusercontent.com/57290520/114273673-96e9dc00-9a4d-11eb-95d2-6627e1a50c2d.jpg)
> and I also try to run the code twice by only change the workflow,the result seems much normal.
> ![kkk](https://user-images.githubusercontent.com/57290520/114273820-2c856b80-9a4e-11eb-8e6f-c870ef6478be.jpg)
> so I am sure about the loss of the val_step output is correct.Looking forward to your replay.

Hi,
Both the graphs seem logical to me. The losses are approaching to zero as the number of epochs are increasing. Let me know if you have any other questions or you need any other help. 
Have a nice day :)
